blog/_src/posts/2017-02-21-bullets-are-good-for-your-coq-proofs.md

```diff
@@ -0,0 +1,239 @@
+
+```
+induction foo. (* this creates many subgoal *)
+
+proof of first subgoal.
+
+proof of second subgoal.
+```
+
+There are many ways to structure this to make the structure more
+apparent: people would typically have a comment on each subgoal, or
+make disciplined use of indentation and blank lines. But, in my
+experience, a major problem with this style was maintainability in the
+face of changes to the definitions or parts of automation. It could be
+very hard of what was happening when a proof suddenly broke after
+a change before in the file:
+
+- If a proof step now proves *less* things, then what used to be the
+  end of a subgoal may not be anymore. Coq would then start reading
+  the proof of the next subgoal and try to apply it to the unfinished
+  previous goals, generating very confusing errors (you believe you
+  are in the second subgoal, but the context talks about a leaf case
+  of the first goal).
+
+- If a proof step now proves *more* things, it is also very bad! The
+  next proof steps, meant for the first subgoal (for example), would
+  then apply to the beginning of the second subgoal, and you get very
+  confusing errors again.
+
+What we need for robustness is a way to indicate our *intent* to Coq,
+when we think that a subgoal is finished and that a new subgoal
+starts, so that Coq can fail loudly at the moment where it notices
+that this intent does not match reality, instead of at an arbitrary
+later time.
+
+(The `S*Case` tactics used in (older versions of) Software Foundations
+can solve this problem if used in a carefully, systematic way, and
+additionally provides naming. Alexandre Pilkiewicz implemented an even
+more powerful [cases](https://github.com/pilki/cases) plugin. Bullets
+are available in standard Coq since 8.4 (released in 2012), and can be
+used with no effort.)
+
+There is not much discussion of bullets around; see the
+[documentation](https://coq.inria.fr/distrib/8.6/refman/Reference-Manual009.html#sec326)
+in the Coq manual. I learned a lot from Arthur Azevedo de
+Amorim's
+[Bullets.v](https://github.com/arthuraa/poleiro/blob/master/theories/Bullets.v)
+file.
+
+Finally, some people don't use bullets, because they systematically
+use so much automation that they never see subgoals -- each lemma has
+a one-line proof. This is also a valid style. (I have been going to
+Adam Chlipala's [Formal Reasoning about
+Programs](https://frap.csail.mit.edu/main) 2017 class, where Adam
+ignores bullets because that is his usual style.) Because I am not
+crushy enough to do this from the start, my proofs tend to start with
+cases and subgoals, and then I refine them to add more automation for
+robustness. I found bullets very useful for the first step, and during
+the refinement process.
+
+# Bullets
+
+Bullets are actually a combination of two features, braces `{ ... }`
+and actually list bullets, `-`, `+`, or `*` or homogeneous repetitions
+of those, for example `--` or `***`.
+
```

Comment:
This sentence is confusing to me. It says there's 2 features but seemingly mentions 3, plus the "or `*` or" makes it hard to parse. Maybe you meant the word "actual" instead of "actually"?